### ISSUE TITLE
Tracking of book reading

### ISSUE THREAD

**Issue by n00b12345** [sbondCo/Watcharr#382]
Book tracking, i.e. tracking progress of books/read/unread books etc. would be really nice to have. Would you consider adding book reading/tracking as a feature?

Watcharr tracks everything else, books are the only thing missing :)

**Comment by IRHM** [comment]
Hah you are right, I hadn't thought of books. I'm definitely open to adding support.

Just need to find a database with a free api we can use and It should hopefully be straightforward enough to integrate it in alongside everything else.

Hope to get to this soon for you.

**Comment by n00b12345** [comment]
> Hah you are right, I hadn't thought of books. I'm definitely open to adding support.
> 
> Just need to find a database with a free api we can use and It should hopefully be straightforward enough to integrate it in alongside everything else.
> 
> Hope to get to this soon for you.

Thank you so much for your consideration.

"Competing" projects like [Media Tracker](https://github.com/bonukai/MediaTracker) or [Jelu](https://github.com/bayang/jelu) use [Open Library](https://openlibrary.org/developers/api) (free API is available) to fetch metadata related to books. This includes things like title, author, covers, description etc. Although not directly related, [Calibre](https://github.com/kovidgoyal/calibre) and [Calibre-web](https://github.com/janeczku/calibre-web) also use Open Library to fetch all kinds of metadata related to books.

I hope this helps a little :)

**Comment by reitenth** [comment]
i would also like to see manga which u could get from anilist/myanimelist

**Comment by reitenth** [comment]
any update on manga support?

**Comment by IRHM** [comment]
@reitenth I haven't started work on this yet, if nothing pressing comes up, I will likely try working on this for the next version (after current one in dev is released).

One thing I will have to take into consideration is how OpenLibrary and AniList (I don't think MAL has an api to access manga data) data will work together.

It seems like OpenLibrary does have manga, which I think would be confusing when mixed with AniList manga in the same results (when searching, or possibly in your watched list).

The only solution I can think of right now for search, would be to never combine books/manga in the main (combined) search and only allow search for Books (OpenLibrary) or Manga (AniList) separately, with the search filters.

I will likely implement both together so that I can make sure they work nicely in the same world.

**Comment by n00b12345** [comment]
I wanted to add a few details so that it might help. A few days back I built a tool for tracking books using ChatGPT. This is what it looks like. It uses PostgreSQL tables in the backend and in being served from on a basic nginx server.

This is the welcome page (nothing is editable here):

![Screenshot 2024-10-13 123413](https://github.com/user-attachments/assets/19e6e370-4af7-42f3-9d51-2e6b609c9821)

This is the "My Books" page which lists all books (including books which have been read and which are yet to be read):

![Screenshot 2024-10-13 123427](https://github.com/user-attachments/assets/3fc34cfe-81a6-4251-a125-e68cda0e3fc9)

The My Books page also has a drop down menu which can sort books:

![Screenshot 2024-10-13 123445](https://github.com/user-attachments/assets/71a5a2f6-4605-4049-98e2-dc361294329f)

This is the To Read section:

![Screenshot 2024-10-13 123502](https://github.com/user-attachments/assets/aece6ed0-c909-40e9-b15f-94785cf7b414)

This is the Add Books section (I have not used any API and this is very basic):

![Screenshot 2024-10-13 123516](https://github.com/user-attachments/assets/2daf84c1-1d3c-4a26-9b37-84b4b7c577a9)

The options in the form change when you add a book that you have already read.

![Screenshot 2024-10-13 123532](https://github.com/user-attachments/assets/0615e582-b7b8-4262-9259-018a937bef1f)

This is the stats section (you can sort books that you have read by the year):

![Screenshot 2024-10-13 123549](https://github.com/user-attachments/assets/3154a134-6aae-4e92-a97c-855f8903e2c8)

This is the search page:

![Screenshot 2024-10-13 123612](https://github.com/user-attachments/assets/1be50593-3659-4b3f-8bd1-1313db9e5374)

This is the progress page where you can set goals and set reading progress which is then reflected on the welcome page:

![Screenshot 2024-10-13 123622](https://github.com/user-attachments/assets/6fe7dcc9-a666-47ef-9e72-b56c04b8183d)

When you tap on a book cover anywhere, this is the viewing page that you are taken to:

![Screenshot 2024-10-13 123636](https://github.com/user-attachments/assets/80537c1a-2af3-4f9f-b5f8-fa4df5b03a36)

The names of the author and tags are all links that then open to their respective pages:

![Screenshot 2024-10-13 123658](https://github.com/user-attachments/assets/a1a9d3b1-2f9c-462b-b4f0-41f11324af9d)

![Screenshot 2024-10-13 123712](https://github.com/user-attachments/assets/37e94c5b-44e3-4b3e-ae4b-29355deeb9fb)

When you tap on edit, this is the editing option that you get:

![Screenshot 2024-10-13 123725](https://github.com/user-attachments/assets/52e2b512-c5cf-4360-9575-cc0d1cc1d64f)

The options change based on the reading status:

![Screenshot 2024-10-13 123738](https://github.com/user-attachments/assets/05581281-f565-41ff-a26f-658efe4f4cbd)

These are the optimised pages for mobile:

![Screenshot_20241013_125246_Firefox](https://github.com/user-attachments/assets/3ab1b5f4-5c8b-4b7e-a4e1-b2f262cd6429)

![Screenshot_20241013_125313_Firefox](https://github.com/user-attachments/assets/e5941a2b-edfe-4acd-a340-ff0bcb69e99b)


I am not a developer and I only generated these using ChatGPT, but I hope some of this helps. If not, sorry :)

**Comment by LamahHerder** [comment]
Hope you have a chance to consider this again
ebooks, audiobooks, comics, manga have had a long history of unfriendly metadata sources
All of these types would be great to start tracking, but they are all different

For the ebooks and audiobooks
Couple more recent options to help with this
self-hosted option https://github.com/blampe/rreading-glasses
public metadata https://docs.hardcover.app/api/getting-started/

**Comment by Bnyro** [comment]
Hey IRHM, I was looking into adding book support via https://openlibrary.org, however I noticed that you're currently working on a large refactor (pagination, ...) on the `pagination-2` branch.

Hence my question: Do you think that would conflict with the required changes for adding some basic book support (i.e. search, book details, saving/rating books), so that it makes sense to wait until you're done refactoring?

PS: Adding book support is probably going to take multiple days of work, so I don't want to risk any conflicts or other reasons from it not being merged in the end.

**Comment by IRHM** [comment]
Thanks for checking in @Bnyro, every file in the backend has moved and been modified quite a bit so any changes would best be made after I can get 3.0 out.

On the bright side, the refactor should make your coding experience more pleasant.

When you get started, feel free to reply again and I can assign you, we can also try going over a basic overview of how it could be implemented

**Comment by Bnyro** [comment]
> On the bright side, the refactor should make your coding experience more pleasant.
>
Sounds great :)

> When you get started, feel free to reply again and I can assign you, we can also try going over a basic overview of how it could be implemented
>
Yeah, that would be very helpful. I've already got a basic idea about how I would implement it with the current source code, but I haven't yet looked into your WIP changes. Anyways it definitely makes sense to have an agreement before wasting a lot of time refactoring things again and again on my end :). I have already made that mistake previously, can't recommend it...

So I'd be very thankful to get a heads-up round-about when you think it makes sense to start working on book support once 3.0 is out/close enough, so that I can start working on it (if my time schedule allow me to :))

**Comment by Bnyro** [comment]
Hey @IRHM, now that v3 is here (🎉 ), in theory we should be ready to start with the book tracking support now? I'm currently a bit busy, but I'd probably find time to start working on this by the end of next week :)

**Comment by IRHM** [comment]
Hi @Bnyro, yes, if you can find the time, the server code is stable enough that I don't expect any big changes in the near future, so it's ready!

And no rush/worries if you have other obligations and can't come back to this or can only get a bit done, I have a lot of time to help out right now or complete things.

---

I know you said you'd add support for OpenLibrary, but I wonder if we need to consider how other book apis might integrate in the future, just so it isn't a pain down the line when we need Google Books support or something else for manga, etc.

And.. I just realized I have an old comment in this thread with the same fears, I can't think of a good solution other than keeping them separate and if there's dupes, whatever?

Wonder if you or anyone else has any thoughts on that?

**Comment by IRHM** [comment]
I just realized we'd have to support Authors too, which clash with Actors, etc.

I think we should be able to reuse the pages etc, but just somehow make sure we get data from the requested places.

**Comment by Bnyro** [comment]
> I know you said you'd add support for OpenLibrary, but I wonder if we need to consider how other book apis might integrate in the future, just so it isn't a pain down the line when we need Google Books support or something else for manga, etc.
>
Yeah, I'll try to keep the data structure for storing books as simple as possible. I'll probably just build an interface with methods like `Search`, `GetBook`, `GetAuthor` so that other providers should be easy to implement in the future.

**Comment by Bnyro** [comment]
I think it'd be great if we can re-use most of the existing movies UI, because otherwise the new pages would probably be 99% copy and paste from the existing pages with only small changes anyways. I.e. book metadata is very similar to movie metadata in many aspects.

**Comment by Bnyro** [comment]
If we additionally ever want to support Google Books, here are the relevant API paths:
- Search books: https://www.googleapis.com/books/v1/volumes?q=test
- Get book details: https://www.googleapis.com/books/v1/volumes/P5skEQAAQBAJ
- Get books by author: https://www.googleapis.com/books/v1/volumes?q=inauthor:"Dan Brown"
- Getting Author details: not supported
